overlord/snapstate/target.go

```diff
@@ -0,0 +1,569 @@
+}
+
+// storeInstallGoal implements the Target interface and represents a group of
+// snaps that are to be installed from the store.
+type storeInstallGoal struct {
+	snaps map[string]StoreSnap
+}
+
+// StoreSnap represents a snap that is to be installed from the store.
+type StoreSnap struct {
+	// InstanceName is the name of snap to install.
+	InstanceName string
+	// Components is the list of components to install with this snap.
+	Components []string
+	// RevOpts contains options that apply to the installation of this snap.
+	RevOpts RevisionOptions
+	// SkipIfPresent indicates that the snap should not be installed if it is already present.
+	SkipIfPresent bool
+}
+
+// verify that StoreTarget implements the Target interface
+var _ InstallGoal = &storeInstallGoal{}
+
+// StoreGoal creates a new storeInstallGoal, which implements InstallGoal to
+// install snaps from the store.
+func StoreGoal(snaps ...StoreSnap) *storeInstallGoal {
+	mapping := make(map[string]StoreSnap, len(snaps))
+	for _, sn := range snaps {
+		if _, ok := mapping[sn.InstanceName]; ok {
+			continue
+		}
+
+		if sn.RevOpts.Channel == "" {
+			sn.RevOpts.Channel = "stable"
+		}
+
+		mapping[sn.InstanceName] = sn
+	}
+
+	return &storeInstallGoal{
+		snaps: mapping,
+	}
+}
+
+func validateRevisionOpts(opts RevisionOptions) error {
+	if opts.CohortKey != "" && opts.Revision.Set() {
+		return errors.New("cannot specify revision and cohort")
+	}
+
+	return nil
+}
+
+// Install returns the data needed to setup the snaps from the store for
+// installation.
+func (s *storeInstallGoal) Install(ctx context.Context, st *state.State, opts Options) ([]Target, error) {
+	allSnaps, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.validateAndPrune(allSnaps); err != nil {
+		return nil, err
+	}
+
+	// create a closure that will lazily load the enforced validation sets if
+	// any of the targets require them
+	var vsets *snapasserts.ValidationSets
+	enforcedSets := func() (*snapasserts.ValidationSets, error) {
+		if vsets != nil {
+			return vsets, nil
+		}
+
+		var err error
+		vsets, err = EnforcedValidationSets(st)
+		if err != nil {
+			return nil, err
+		}
+
+		return vsets, nil
+	}
+
+	actions := make([]*store.SnapAction, 0, len(s.snaps))
+	for _, sn := range s.snaps {
+		action, err := installActionForStoreTarget(sn, opts, enforcedSets)
+		if err != nil {
+			return nil, err
+		}
+
+		actions = append(actions, action)
+	}
+
+	curSnaps, err := currentSnaps(st)
+	if err != nil {
+		return nil, err
+	}
+
+	refreshOpts, err := refreshOptions(st, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	user, err := userFromUserID(st, opts.UserID)
+	if err != nil {
+		return nil, err
+	}
+
+	str := Store(st, opts.DeviceCtx)
+
+	st.Unlock() // calls to the store should be done without holding the state lock
+	defer st.Lock()
+
+	results, _, err := str.SnapAction(context.TODO(), curSnaps, actions, nil, user, refreshOpts)
+	if err != nil {
+		if len(actions) == 1 {
```

Comment:
I'm not sure this is quite the right approach, passing 1 snap to a Many function previously has different error behavior than passing it to a single target function. We might need to have a Single flag in options or something to control this better

Reply:
Added in 938cd26c6694412ea13e8bb443e0c164b3aa6e1f.